### ISSUE TITLE
matmul with CSR matrix in inference mode throws an exception

### ISSUE THREAD

**Issue by KrisDemuynck** [pytorch/pytorch#98004]
### 🐛 Describe the bug

```
# BUG: matmul with CSR matrix in inference mode throws an exception
import torch;
x = torch.randn((4,8,5),dtype=torch.float);
M = torch.Tensor.to_sparse_csr(torch.eye(8));
# non-inference mode works
y = torch.matmul(M,x);
# inference mode throws RuntimeError: Cannot set version_counter for inference tensor
with torch.inference_mode():
     y = torch.matmul(M,x);
```

### Versions

Collecting environment information...
PyTorch version: 1.13.1+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: 10.0.0-4ubuntu1 
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.8.10 (default, Nov 14 2022, 12:59:47)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.4.0-137-generic-x86_64-with-glibc2.29
Is CUDA available: True
CUDA runtime version: 12.0.76
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA TITAN Xp
Nvidia driver version: 525.60.13
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.7.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   39 bits physical, 48 bits virtual
CPU(s):                          12
On-line CPU(s) list:             0-11
Thread(s) per core:              2
Core(s) per socket:              6
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           158
Model name:                      Intel(R) Core(TM) i7-8700K CPU @ 3.70GHz
Stepping:                        10
CPU MHz:                         4300.000
CPU max MHz:                     4700.0000
CPU min MHz:                     800.0000
BogoMIPS:                        7399.70
Virtualization:                  VT-x
L1d cache:                       192 KiB
L1i cache:                       192 KiB
L2 cache:                        1.5 MiB
L3 cache:                        12 MiB
NUMA node0 CPU(s):               0-11
Vulnerability Itlb multihit:     KVM: Vulnerable
Vulnerability L1tf:              Mitigation; PTE Inversion
Vulnerability Mds:               Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Meltdown:          Mitigation; PTI
Vulnerability Mmio stale data:   Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:          Mitigation; IBRS
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Mitigation; Microcode
Vulnerability Tsx async abort:   Mitigation; Clear CPU buffers; SMT vulnerable
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp md_clear flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.21.1
[pip3] numpydoc==0.7.0
[pip3] torch==1.13.1
[pip3] torchaudio==0.13.1
[pip3] torchvision==0.14.1
[conda] Could not collect

cc @alexsamardzic @nikitaved @pearu @cpuhrsch @amjames @bhosmer

**Comment by pat749** [comment]
A workaround for this issue is to avoid using `torch.inference_mode()` when working with sparse CSR matrices, or to convert them to dense tensors before using them in the operation.
```
import torch
x = torch.randn((4, 8, 5), dtype=torch.float)
M = torch.Tensor.to_sparse_csr(torch.eye(8))

# convert to dense tensor before using matmul
M_dense = M.to_dense()
with torch.inference_mode():
    y = torch.matmul(M_dense, x)
```
This changes will not throw error.

And also This is a known bug in PyTorch when using `torch.inference_mode()` with sparse CSR matrices. Inference mode is meant for faster and more memory-efficient inference computations by disabling some internal checks and features. However, for some operations with sparse tensors, such as `torch.matmul`, PyTorch needs to update internal version counters for correctness, and this is not allowed in inference mode. As a result, you may encounter a RuntimeError with the message "Cannot set version_counter for inference tensor" when running such operations.

You can read more here in this documentation of [pytorch](https://pytorch.org/docs/stable/sparse.html).

**Comment by KrisDemuynck** [comment]
Thanks for the quick and informative reply!

Given that my matrix is huge (4045x4045), very sparse (0.6% filled) and needs to be applied billions of times, disabling the inference mode is the best option (a solution I was already testing and which seems to work well; However this still felt like a "hack", hence the "bug" report).

As for the "This is a known bug" ...

- I did use Google to search for the "Cannot set version_counter for inference tensor" message, but found zero useful pages (except for the github page with the C-code, which led me to try disabling the inference mode).
- The pytorch sparse documentation also does not mention the word "inference" at all ... (but does state the beta state of the sparse matrix support, and asks to report problems; which I diligently did ...

Note: Adding your clear explanation to the pytorch sparse documentation would have been very useful for me.

**Comment by pat749** [comment]
@KrisDemuynck I will pull request to add this. Thank you

**Comment by malfet** [comment]
Hmm,

**Comment by cpuhrsch** [comment]
@pearu - can you take a look?

**Comment by malfet** [comment]
> Note that for me, on latest master, the first matmul fails with `RuntimeError: Sparse CSR tensors do not have strides`.

Yeah, I'm seeing the same, but I've assumed, that this is because I build without MKL support, though on M1 everything works fine

**Comment by malfet** [comment]
> @pearu - can you take a look? Looks like we might need a smoke test for this.

@cpuhrsch come on, let me have some fun with it :)

**Comment by pearu** [comment]
> Note that for me, on latest master, the first matmul fails with `RuntimeError: Sparse CSR tensors do not have strides`.

This typically means that backward machinery has kicked in and calling detach often helps.

**Comment by pearu** [comment]
@malfet not really an answer but here's hopefully a useful hint:
```python
>>> x = torch.randn((4,8,5),dtype=torch.float)
>>> M = torch.Tensor.to_sparse_csr(torch.eye(8)).requires_grad_(False)
>>> y = torch.matmul(M,x);
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Sparse CSR tensors do not have strides
>>> x = torch.randn((4,8,5),dtype=torch.float)
>>> M = torch.Tensor.to_sparse_csr(torch.eye(8)).requires_grad_(True)
>>> y = torch.matmul(M,x);
```

**Comment by malfet** [comment]
Looks like above regressed sometime between Mar 30st and Mar 31st nightlies
